src/platform/K32W/KeyValueStoreManagerImpl.cpp

```diff
@@ -24,76 +24,189 @@
 #include <platform/internal/CHIPDeviceLayerInternal.h>
 
 #include <lib/support/CHIPMem.h>
+#include <list>
 #include <platform/K32W/K32WConfig.h>
 #include <platform/KeyValueStoreManager.h>
 #include <string>
 
 #include "PDM.h"
 
+#include <unordered_map>
+
 namespace chip {
 namespace DeviceLayer {
 namespace PersistedStorage {
 
 /* TODO: adjust this value */
-#define MAX_NO_OF_KEYS 255
+#define MAX_NO_OF_KEYS 20
 
 KeyValueStoreManagerImpl KeyValueStoreManagerImpl::sInstance;
 
-CHIP_ERROR KeyValueStoreManagerImpl::_Get(const char * key, void * value, size_t value_size, size_t * read_bytes_size,
-                                          size_t offset_bytes)
+/* hashmap having:
+ * 	- the matter key string as key;
+ * 	- internal PDM identifier as value;
+ */
+std::unordered_map<std::string, uint8_t> g_kvs_map;
+
+/* list containing used PDM identifiers */
+std::list<uint8_t> g_key_ids_list;
+
+/* max no of bytes for a key */
+#define MAX_KEY_VALUE          255
+
+/* used to check if we need to restore values from flash (e.g.: reset) */
+static bool g_restored_from_flash = FALSE;
+
+CHIP_ERROR RestoreFromFlash()
 {
     CHIP_ERROR err = CHIP_NO_ERROR;
-    std::hash<std::string> hash_fn;
-    uint16_t key_id;
+    uint8_t key_id = 0;
+    char key_string_id[MAX_KEY_VALUE] = {0};
+    size_t key_string_id_size = 0;
     uint8_t pdm_id_kvs = chip::DeviceLayer::Internal::K32WConfig::kPDMId_KVS;
-    size_t read_bytes;
 
-    VerifyOrExit((key != NULL) && (value != NULL), err = CHIP_ERROR_INVALID_ARGUMENT);
-    key_id = hash_fn(key) % MAX_NO_OF_KEYS;
+    if (g_restored_from_flash)
+    {
+        /* already restored from flash, nothing to do */
+        return err;
+    }
 
-    ChipLogProgress(DeviceLayer, "KVS, get key id:: %i", key_id);
+    for (key_id = 0; key_id < MAX_NO_OF_KEYS; key_id++)
+    {
+        /* key was saved as string in flash (key_string_id) using (key_id + MAX_NO_OF_KEYS) as PDM key
+         * value corresponding to key_string_id was saved in flash using key_id as PDM key
+         */
 
-    err = chip::DeviceLayer::Internal::K32WConfig::ReadConfigValueBin(
-        chip::DeviceLayer::Internal::K32WConfigKey(pdm_id_kvs, key_id), (uint8_t *) value, value_size, read_bytes);
+        err = chip::DeviceLayer::Internal::K32WConfig::ReadConfigValueStr(
```

Comment:
this seems racy: if the last read is success, then this method will return succes. if the last is NOT_FOUND, the method returns NOT_FOUND.

Either consistently clear the error or return error on first failure. It sounds like the logic should be:

- if NOT_FOUND then set error to NO_ERROR and continue
- on any error except NO_ERROR or NOT_FOUND, return an error

Reply:
Thanks, use the proposed logic.

Reply:
> Either consistently clear the error or return error on first failure. It sounds like the logic should be:
> 
>     * if NOT_FOUND then set error to NO_ERROR and continue
> 
>     * on any error except NO_ERROR or NOT_FOUND, return an error

Current code sets g_restored_from_flash when returning the error.

I think this is not correct.  This means that subsequent ops will silently ignore the fact that initialization has actually failed.

Reply:
Added suggestion here: https://github.com/project-chip/connectedhomeip/pull/9489/files#r744130070.